File: 637/average-of-levels-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
function averageOfLevels(root) {
    let nodes = [root];
    let res = [];
    while (nodes.length > 0) {
        [nodes, avg] = level_nodes_and_avg(nodes);
        res.push(avg);
    }
    return res;
}

function level_nodes_and_avg(nodes) {
    let next_nodes = [];
    let avg = 0;
    for (let node of nodes) {
        avg += node.val;
        if (node.left)
            next_nodes.push(node.left);
        if (node.right)
            next_nodes.push(node.right);
    }
    avg /= nodes.length;
    return [next_nodes, avg];
}

File: 637/average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        nodes = [root]
        res = []
        while len(nodes) > 0:
            nodes, avg = self.level_nodes_and_avg(nodes)
            res.append(avg)
        return res

    def level_nodes_and_avg(self, nodes):
        """
        :type nodes: List[TreeNode]
        :rtype next_nodes, avg: List[TreeNode], float
        """
        next_nodes = []
        avg = 0
        for node in nodes:
            avg += node.val
            if node.left:
                next_nodes.append(node.left)
            if node.right:
                next_nodes.append(node.right)
        avg /= len(nodes)
        return next_nodes, avg
